unpack_unix_test.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:build unix
+
+package extract_test
+
+import (
+	"context"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-extract"
+)
+
+func TestUnpackWithPreserveFileAttributes(t *testing.T) {
+	uid, gid := 503, 20
+	baseTime := time.Date(2021, 1, 1, 0, 0, 0, 0, time.Local)
+	testCases := []struct {
+		name                  string
+		contents              []archiveContent
+		packer                func(*testing.T, []archiveContent) []byte
+		doesNotSupportModTime bool
+		doesNotSupportOwner   bool
+		expectError           bool
+	}{
+		{
+			name: "tar",
+			contents: []archiveContent{
+				{Name: "test", Content: []byte("hello world"), Mode: 0777, AccessTime: baseTime, ModTime: baseTime, Uid: 0, Gid: 0},
+				{Name: "sub", Mode: fs.ModeDir | 0777, AccessTime: baseTime, ModTime: baseTime, Uid: uid, Gid: gid},
+				{Name: "sub/test", Content: []byte("hello world"), Mode: 0777, AccessTime: baseTime, ModTime: baseTime, Uid: uid, Gid: gid},
+				{Name: "link", Mode: fs.ModeSymlink | 0777, Linktarget: "sub/test", AccessTime: baseTime, ModTime: baseTime},
+			},
+			packer: packTar,
+		},
+		{
+			name: "zip",
+			contents: []archiveContent{
+				{Name: "test", Content: []byte("hello world"), Mode: 0777, AccessTime: baseTime, ModTime: baseTime, Uid: uid, Gid: gid},
+				{Name: "sub", Mode: fs.ModeDir | 0777, AccessTime: baseTime, ModTime: baseTime, Uid: uid, Gid: gid},
+				{Name: "sub/test", Content: []byte("hello world"), Mode: 0644, AccessTime: baseTime, ModTime: baseTime, Uid: uid, Gid: gid},
+				{Name: "link", Mode: fs.ModeSymlink | 0777, Linktarget: "sub/test", AccessTime: baseTime, ModTime: baseTime},
+			},
+			doesNotSupportOwner: true,
+			packer:              packZip,
+		},
+		{
+			name:                  "rar",
+			contents:              contentsRar2,
+			packer:                packRar2,
+			doesNotSupportModTime: true,
+		},
+		{
+			name:                "7z",
+			contents:            contents7z2,
+			doesNotSupportOwner: true,
+			packer:              pack7z2,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				ctx = context.Background()
+				dst = t.TempDir()
+				src = asIoReader(t, tc.packer(t, tc.contents))
+				cfg = extract.NewConfig(extract.WithPreserveFileAttributes(true))
+			)
+			if err := extract.Unpack(ctx, dst, src, cfg); err != nil {
+				t.Fatalf("error unpacking archive: %v", err)
+			}
+			for _, c := range tc.contents {
+				path := filepath.Join(dst, c.Name)
+				stat, err := os.Lstat(path)
+				if err != nil {
+					t.Fatalf("error getting file stats: %v", err)
+				}
+				if !(c.Mode&fs.ModeSymlink != 0) { // skip symlink checks
+					if stat.Mode().Perm() != c.Mode.Perm() {
+						t.Fatalf("expected file mode %v, got %v, file %s", c.Mode.Perm(), stat.Mode().Perm(), c.Name)
+					}
+				}
+				if tc.doesNotSupportModTime {
+					continue
+				}
```

Comment:
Because of this check, we are also skipping the UID/GID check for `rar` archives. I found that the UID/GID check is failing for `rar` files when we do run the assertion... See #111

Reply:
Ohhhhhh, you are right! Great catch!

Reply:
Dang, I missed to add a flag that rar test case so that `tc.doesNotSupportOwner = true` is set as well. 

Looking in the `Uid`/`Gid` function reveals that the ownership is not extracted from the rar archive, bc/ not supported by the library:

https://github.com/hashicorp/go-extract/blob/96659b8a432b44ec38fa45e0ff00407466da08ce/rar.go#L155-L163

(I recognised right now that I missed to add a propper code comment for that.)

Reply:
I adjusted the code comment in https://github.com/hashicorp/go-extract/pull/106/commits/f4bfff1ffd2b2c289e99bdccc7e29efd18aadb06.

Reply:
As by https://github.com/hashicorp/go-extract/pull/106#discussion_r1886599366 we have now two flags to cover that and separate test cases :)